bool train
bool save
int32 step
---
bool result